rfd/0173-terraform-machine-id.md

```diff
@@ -0,0 +1,303 @@
+to production, they will hopefully turn into [existing advanced users](#the-existing-advanced-user) and setup CI/CD
+pipelines for their deployment.
+
+### Implementation
+
+We will introduce two new mechanisms in the Terraform provider:
+- MachineID joining
+- Resource bootstrapping
+
+#### MachineID joining
+
+The Terraform provider will be able to take a MachineID token and join method to join the Teleport cluster without the
+need to configure and run a sidecar tbot process.
+
+The Terraform provider will run an in-process `tbot` instance and generate a client with automatic cert-renewal.
+The `tbot` instance will not persist its certificates on the disk.
+
+A typical GitHub actions provider configuration would look like:
+
+```hcl
+provider "teleport" {
+  addr               = "mytenant.teleport.sh:443"
+  onboarding = {
+    token       = "gha-runner"
+    join_method = "github"
+  }
+}
+```
+
+As Teleport does not support yet reusable bot tokens so the `token` join method will not be supported.
+
+Implementation-wise, this would reuse the
+same [in-process tbot library](https://github.com/gravitational/teleport/tree/rfd/173-terraform-machine-id/integrations/operator/embeddedtbot)
+we used for the Teleport operator.
+
+#### Resource bootstrapping
+
+When bootstrapping is turned on, the Terraform provider will use the user's local profile (from `~/.tsh`) to create
+resources it needs to interact with the Teleport cluster. Such resources include:
+
+- the `terraform-provider` role (this is an upsert so we can add new rules after a provider update). This resource does
+  not expire.
+- the `terraform-provider-<user>-<random hash>` bot allowed to issue certificates for the `terraform-provider` role.
+  This resource expires.
+- the `terraform-provider-<user>-<random hash>` secret random provision token allowing to join as
+  the `terraform-provider-<user>-<random hash>` bot. This resource expires but will be consumed automatically on join.
+
+Each local Terraform invocation will create a new bot and join token. To avoid resource accumulation those resource will
+have an expiry of 1 hour by default.
+
+A typical "getting started" provider configuration would look like:
+
+```hcl
+provider "teleport" {
+  addr = "mytenant.teleport.sh:443"
+  bootstrap = {
+    enabled = true
+    resource_prefix = "terraform-provider"
+    
+    # optional, must be higer than your Terraform `apply` command duration
+    expires_after = "1h"
+  }
+}
+```
+
+When running with this configuration, the provider will:
+- get a client from the local `~/.tsh` profile associated with this `addr`
+- ping the cluster to validate the user client and recover the user's name
+- generate a random secret token (16 bytes of hex-encoded random, the Teleport default)
+- create the 3 bootstrap resources
+  - if the call fails because of MFA4A, perform the MFA challenge once, save the MFA certs valid for a minute and use
```

Comment:
MFA4A doesn't issue MFA verified certs, it expects the MFA challenge response to be passed with the API request directly. Since you're making 3 API calls, you'll want to:
1. Check if MFA4A is required
2. Create a MFA challenge that can be reused for all 3 API calls (reuse challenge extension)
3. Prompt the user to answer the MFA challenge
4. Attach the MFA challenge response in the `ctx` for each API call.

See https://github.com/gravitational/teleport/pull/37121 for an example of this flow

Reply:
Updated the RFD based on your comment